haystack/dataclasses/tool.py

```diff
@@ -0,0 +1,18 @@
+from typing import Any, Callable, Dict
+
+
+class Tool:
```

Comment:
this is very minimal and currently contains no methods to automatically convert existing functions to tools.

I'm not sure if it should go into experimental:
- it is experimental but...
- the Chat Generators in Haystack will need it (for the `run` method)
- the Chat Generators in core-integrations will need it (for the `run` method)

Let's discuss this...

Reply:
The tool-related changes and additions will go into experimental. I'd keep the list of supported chat generators to a minimum for the duration of the experiment, i.e., let's pick three. This is how I see it:
- Tool-related dataclasses, interfaces are added to the experimental package.
- For generators that are part of the core package, we just override them in experimental without any extra ceremony.
- For generators that are part of a core integration, we lazy-import them into experimental (like we do with the OpenAPITool) and override them.

Reply:
I understand the idea and we can do it.

In general, this will require a communication effort, I think.

For example, will we have to tell people that to import Anthropic with this functionality they will have to do `from haystack_experimental.components.generators imprort AnthropicChatGenerator`?

Will the new `run` method tests with the tools have to be on `haystack_experimental`? Will we have to partially reproduce the CI of core integrations there?

Reply:
> For example, will we have to tell people that to import Anthropic with this functionality they will have to do from haystack_experimental.components.generators imprort AnthropicChatGenerator?

Yeah, basically. But as with the comms for the other available experiments, we'd mention this a part of the example notebook.

> Will the new run method tests with the tools have to be on haystack_experimental? Will we have to partially reproduce the CI of core integrations there?

Yes, the tests will be written in such a way that they can be ported with minimal effort into core/core integrations at the end of the experiment. We don't need to reproduce all the tests - Just the ones that will be affected the new changes. 

I don't think we need to reproduce the CI of core integrations in the experimental repo. Its CI already has secrets for some LLM providers (added as part of the OpenAPI tool PR) - the new tests will just run for all PRs.